nx/lib/nx/binary_backend/traverser.ex

```diff
@@ -0,0 +1,269 @@
+      readers_template: readers,
+      readers: readers
+    }
+  end
+
+  defp do_build_paths(shape, [_ | _] = axes, limits, dilations, transpose) do
+    axes = Enum.sort(axes)
+    min = hd(axes)
+
+    shape
+    |> weighted_shape(limits, dilations)
+    |> transpose_weighted_shape(transpose)
+    |> Enum.drop(min)
+    |> aggregate_paths(axes, min, [], [])
+  end
+
+  defp do_build_paths(shape, [], limits, dilations, transpose) do
+    shape
+    |> weighted_shape(limits, dilations)
+    |> transpose_weighted_shape(transpose)
+    |> aggregate_paths([], 0, [], [])
+  end
+
+  def next(%Traverser{offsets: []} = trav) do
+    case next_cycle(trav) do
+      :done ->
+        :done
+
+      trav ->
+        next(trav)
+    end
+  end
+
+  def next(%Traverser{offsets: [_ | offsets], readers: [], readers_template: rt} = trav) do
+    next(%Traverser{trav | offsets: offsets, readers: rt})
+  end
+
+  def next(%Traverser{offsets: [o | _], readers: readers} = trav) do
+    {:cont, r, rest} = next_item(readers)
+    {:cont, o + r, %Traverser{trav | readers: rest}}
+  end
+
+  def next_agg(
+        %Traverser{
+          offsets: [head_o | offsets],
+          readers: readers,
+          readers_template: readers_template
+        } = trav
+      ) do
+    size = length(readers_template)
+    agg = do_build_struct(size, size, [head_o], readers)
+    trav = %Traverser{trav | offsets: offsets, readers: readers_template}
+    {:cont, agg, trav}
+  end
+
+  def next_agg(%Traverser{offsets: []} = trav) do
+    case next_cycle(trav) do
+      :done ->
+        :done
+
+      trav ->
+        next_agg(trav)
+    end
+  end
+
+  def agg_iter(trav) do
+    AggIter.build(trav)
+  end
+
+  defimpl Enumerable do
+    def count(%Traverser{size: size}) do
+      {:ok, size}
+    end
+
+    def member?(_, _) do
+      {:error, __MODULE__}
+    end
+
+    def slice(_) do
+      {:error, __MODULE__}
+    end
+
+    def reduce(_, {:halt, acc}, _fun) do
+      {:halted, acc}
+    end
+
+    def reduce(trav, {:suspend, acc}, fun) do
+      {:suspended, acc, fn acc2 -> reduce(trav, acc2, fun) end}
+    end
+
+    def reduce(trav, {:cont, acc}, fun) do
+      case Traverser.next(trav) do
+        {:cont, i, trav2} ->
+          reduce(trav2, fun.(i, acc), fun)
+
+        :done ->
+          {:done, acc}
+      end
+    end
+  end
+
+  defp next_cycle(trav) do
+    %Traverser{
+      cycle: cycle_prev,
+      size: size,
+      cycle_size: cycle_size,
+      offsets_template: offsets_template,
+      readers_template: readers_template
+    } = trav
+
+    cycle = cycle_prev + 1
+
+    if div(size, cycle_size) == cycle do
+      :done
+    else
+      cycle_offset = cycle * cycle_size
+      offsets = Enum.map(offsets_template, fn o -> o + cycle_offset end)
+      %Traverser{trav | cycle: cycle, offsets: offsets, readers: readers_template}
+    end
+  end
+
+  defp next_item([]) do
+    :done
+  end
+
+  defp next_item(start..stop) when start == stop do
```

Comment:
The ranges here are relic of some experimentation. Their use might still be viable.